### PR TITLE
Add Dockerfile

### PR DESCRIPTION
**Pull request by mattias-p**
## Purpose

This PR provides dockerization of Zonemaster Engine.

## Context

This is a step towards zonemaster/zonemaster-cli#216.
Depends on zonemaster/zonemaster-ldns#124.

## Changes

This PR adds a docker file and steps to build it.

## How to test this PR

Follow the new steps in the README file.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,6 +26,29 @@ Installation instructions for the Engine is provided in the
...
+
+```sh
+make docker-tag-version
+```
+
+Tag the local base image as the latest version:
+
+```sh
+make docker-tag-version
+```
```
[the remark this answers is not in the document]
Comment (mattias-p) [reply README.md]: Fixed.

File: Makefile.PL [hunk 2]
```diff
@@ -51,17 +51,31 @@ recommends 'Readonly::XS' => 0;
...
+    my $docker = <<'END_DOCKER';
+
+docker-build: dist
+	sudo docker build --tag zonemaster/engine:local --build-arg version=$(VERSION) .
+
+docker-tag-version: docker-image
+	sudo docker tag zonemaster/engine:local zonemaster/engine:$(VERSION)
+
+docker-tag-latest: docker-image
+	sudo docker tag zonemaster/engine:local zonemaster/engine:latest
```
[the remark this answers is not in the document]
Comment (mattias-p) [reply Makefile.PL]: Fixed.

File: README.md [hunk 3]
```diff
@@ -26,6 +26,29 @@ Installation instructions for the Engine is provided in the
...
+
```
[the remark this answers is not in the document]
Comment (mattias-p) [reply README.md:57]: Fixed.

**Comment by mattias-p** [comment]
> What do I do wrong?

Nothing. Could you try it again now?

**Comment by matsduf** [comment]
> > What do I do wrong?
> 
> Nothing. Could you try it again now?

It works now with the smoke test you provided and with the updates that you provided.

**Comment by blacksponge** [comment]
I don't know what I did wrong but I just have BASIC tests, listing the tests using the cli only return basic and if I try to run 
```
docker run zonemaster/engine:local perl -MZonemaster::Engine -E 'say join "\n", Zonemaster::Engine->test_module("ADDRESS", "zonemaster.net")'
```
I have an empty result

**Comment by mattias-p** [comment]
We have a file that lists the name of all the modules (modules.txt). It seems this file can end up being empty. I haven't figured out how this happens. But when it does only hardcoded BASIC tests are recognized.

**Comment by matsduf** [comment]
It works for me. I have not seen any case where only Basic tests are run. I have also solved the issue when translation does not work. It is necessary to run "make all" in at least Engine and CLI before "make docker-build".

**Comment by matsduf** [comment]
`make docker-build` creates an image tagged with "local".

If then `make docker-tag-version` is run, a new image is created and tagged with "local" and the version number. The first image is either removed or stays untagged (I have not checked).

If then `make docker-tag-latest` is run, a third image is created and tagged with "local" and "latest". The second image remains, but is now only tagged with the version number.

What you want is to be able to create one image that is tagged with all three tags, or possibly create one image tagged with "local" and version number.

I suggest
* That `make docker-build` creates an image tagged with "local" and version number.
* That `make docker-tag-latest` adds the tag "latest" to the image already tagged "local". I there is no such image, it should just fail.

**Comment by mattias-p** [comment]
Aren't there unnecessary risks if we casually set and move the version tag between different builds? I guess it would be easier to push the wrong build to dockerhub or to change a previously uploaded tag by mistake. I'm more comfortable removing the dependency of docker-tag-version on docker-build. Doing the same for docker-tag-latest would solve the problem you noted there as well.

**Comment by mattias-p** [comment]
I just realized that all of the problems above are the results of dependencies on phony targets. It's much better to put `make all dist docker-build` in the documentation than trying to achieve well-behaved tooling that relies on dependencies on phony make targets.

File: Makefile.PL [hunk 4]
```diff
@@ -51,17 +51,31 @@ recommends 'Readonly::XS' => 0;
...
+        $pure_all = 'pure_all :: share/Makefile' . "\n"
             . "\t" . 'cd share && $(MAKE) all' . "\n";
-    };
-    return $text;
+    }
+
+    my $docker = <<'END_DOCKER';
+
+docker-build:
+	docker build --tag zonemaster/engine:local --build-arg version=$(VERSION) .
```
Comment (matsduf) [inline Makefile.PL:68]:
```
$ make docker-build
docker build --tag zonemaster/engine:local --build-arg version=v4.2.2 .
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post http://%2Fvar%2Frun%2Fdocker.sock/v1.24/build?buildargs=%7B%22version%22%3A%22v4.2.2%22%7D&cachefrom=%5B%5D&cgroupparent=&cpuperiod=0&cpuquota=0&cpusetcpus=&cpusetmems=&cpushares=0&dockerfile=Dockerfile&labels=%7B%7D&memory=0&memswap=0&networkmode=default&rm=1&shmsize=0&t=zonemaster%2Fengine%3Alocal&target=&ulimits=null&version=1: dial unix /var/run/docker.sock: connect: permission denied
make: *** [Makefile:969: docker-build] Error 1
```

Comment (mattias-p) [reply Makefile.PL:68]: Yeah, we should have an instruction to add the user to the `docker` group.

**Comment by mattias-p** [comment]
@matsduf @blacksponge Please review.

**Comment by blacksponge** [comment]
> I don't know what I did wrong but I just have BASIC tests, listing the tests using the cli only return basic and if I try to run

I should have missed something, I tried to rebuild and it work now.

**Review by blacksponge — APPROVED** [review]

**Review by matsduf — APPROVED** [review]
When this is merged I will create a PR with updates.

**Pull request merged**